core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnSimulatedData.java

```diff
@@ -0,0 +1,601 @@
+        }
+
+        public List<ChannelDefinition> getChannels()
+        {
+            return channels;
+        }
+    }
+
+    public static class ChannelDefinition
+    {
+        private final ColumnType columnType;
+        private final int distinctValuesCount;
+        private final int dictionaryPositions;
+        private final int distinctDictionaries;
+
+        public ChannelDefinition(ColumnType columnType, int distinctValuesCount)
+        {
+            this(columnType, distinctValuesCount, -1, -1);
+        }
+
+        public ChannelDefinition(ColumnType columnType, int distinctValuesCount, int distinctDictionaries)
+        {
+            this(columnType, distinctValuesCount, distinctDictionaries, distinctValuesCount);
+        }
+
+        public ChannelDefinition(ColumnType columnType, int distinctValuesCount, int distinctDictionaries, int dictionaryPositions)
+        {
+            this.columnType = requireNonNull(columnType, "columnType is null");
+            this.distinctValuesCount = distinctValuesCount;
+            this.dictionaryPositions = dictionaryPositions;
+            this.distinctDictionaries = distinctDictionaries;
+            checkArgument(dictionaryPositions == -1 || dictionaryPositions >= distinctValuesCount);
+        }
+
+        public ColumnType getColumnType()
+        {
+            return columnType;
+        }
+
+        public int getDistinctValuesCount()
+        {
+            return distinctValuesCount;
+        }
+
+        public Block[] createBlocks(int blockCount, int positionsPerBlock, int channel, double nullChance)
+        {
+            Block[] blocks = new Block[blockCount];
+            if (dictionaryPositions == -1) { // No dictionaries
+                createNonDictionaryBlock(blockCount, positionsPerBlock, channel, nullChance, blocks);
+            }
+            else {
+                createDictionaryBlock(blockCount, positionsPerBlock, channel, nullChance, blocks);
+            }
+            return blocks;
+        }
+
+        private void createDictionaryBlock(int blockCount, int positionsPerBlock, int channel, double nullChance, Block[] blocks)
+        {
+            Random r = new Random(1);
+
+            BlockBuilder allValues = generateValues(channel, dictionaryPositions);
+            Set<Integer> usedValues = nOutOfM(r, distinctValuesCount, dictionaryPositions);
+
+            Block[] dictionaries = new Block[distinctDictionaries];
+            Set<Integer>[] possibleIndexesSet = new Set[distinctDictionaries];
+            // Generate dictionaries and positions from those dictionaries that are actually used
+            for (int i = 0; i < distinctDictionaries; i++) {
+                BlockBuilder dictionaryBuilder = columnType.getType().createBlockBuilder(null, dictionaryPositions);
+                possibleIndexesSet[i] = new HashSet<>();
+                Set<Integer> distinctValues = nOutOfM(r, dictionaryPositions, dictionaryPositions);
```

Comment:
`nOutOfM(r, dictionaryPositions, dictionaryPositions);`
why? You choose `N` out of `N`?